daemon/src/db/cull_old_dlcs.rs

```diff
@@ -1,127 +1,82 @@
-///
-/// This method does not modify the database: it just loads rows and
-/// processes them. The returned values can later be used to update
-/// the contents of the database.
-async fn emptied_dlc_events(
-    conn: &mut PoolConnection<Sqlite>,
-) -> Result<impl Iterator<Item = (i64, EventKind)>> {
-    let rows = sqlx::query!(
-        r#"
-        SELECT
-            id,
-            name,
-            data
-        FROM
-            events
-        WHERE
-            name = $1 OR
-            name = $2
-        ORDER BY
-            created_at DESC
-        LIMIT -1 OFFSET 1
-        "#,
-        model::EventKind::CONTRACT_SETUP_COMPLETED_EVENT,
-        model::EventKind::ROLLOVER_COMPLETED_EVENT,
-    )
-    .fetch_all(&mut *conn)
-    .await?;
-
-    Ok(rows
-        .into_iter()
-        .filter_map(|row| {
-            let id = row.id;
-            let event = match EventKind::from_json(row.name, row.data) {
-                Ok(event) => Some(event),
-                Err(e) => {
-                    tracing::warn!("Failed to deserialize EventKind from JSON: {e:#}");
-                    None
-                }
-            };
-
-            event.map(|event| (id, event))
-        })
-        .filter_map(|(id, event)| {
-            use EventKind::*;
-            let event = match event {
-                ContractSetupCompleted { dlc: Some(_), .. } => {
-                    Some(ContractSetupCompleted { dlc: None })
-                }
-                RolloverCompleted {
-                    dlc: Some(_),
-                    funding_fee,
-                } => Some(RolloverCompleted {
-                    dlc: None,
-                    funding_fee,
-                }),
-                // ignore events that have already been culled
-                ContractSetupCompleted { dlc: None } | RolloverCompleted { dlc: None, .. } => None,
-                // ignore all other events
-                _ => None,
-            };
-
-            event.map(|event| (id, event))
-        }))
-}
+    /// Cull old cfd rollover events which are not needed anymore
+    ///
+    /// We set the `dlc` field of events `RolloverCompleted` to null where the event is not the most
+    /// recent `RolloverCompleted` event.
+    async fn cull_rollover_completed(conn: &mut PoolConnection<Sqlite>) -> Result<()> {
+        let effected_rows = sqlx::query!(
+            r#"
+            update EVENTS
+                set
+                    data = (
+                        select json_set(events.data, '$.dlc', null)
+                        from events as inner_events
+                        where inner_events.id = EVENTS.id 
+                    )
+                where name = 'RolloverCompleted'
+                  and id not in (
+                    select max(id) from EVENTS as inner_events
+                    where inner_events.cfd_id = EVENTS.cfd_id
+                      and inner_events.name = EVENTS.name
+                );
```

Comment:
I'm sure you've tried it, but why doesn't something like

```suggestion
                where name = 'RolloverCompleted' OR name = 'ContractSetupCompleted'
                  and id not in (
                    select max(id) from EVENTS as inner_events
                    where inner_events.cfd_id = EVENTS.cfd_id
                      and inner_events.name = EVENTS.name
                );
```

work?

Reply:
Because if we only have `ContractSetupCompleted` but no `RolloverCompleted` yet we want to keep the dlc data.

Reply:
Have you tried it? My crappy SQL brain compiler seems to think that it could work.

Reply:
I tried, but maybe you can come up with a query that works? 

The second query looks slightly different than the first if you look closely: it checks if the `cfd_id` is  in the list of events where there is at least one `RolloverCompletedEvent`

Reply:
> The second query looks slightly different than the first if you look closely: it checks if the cfd_id is in the list of events where there is at least one RolloverCompletedEvent

True. My idea would be to do something very similar to what you do in the first one, but slightly modified so that it considers all contract-setup/rollover events and only keeps the most recent one (whether it is contract-setup or rollover). But there must be a flaw in how I imagine it working.

> I tried, but maybe you can come up with a query that works?

I invoke the other SQL expert @da-kami if he's curious 😛 And if this already works we can iterate on it later.